src/tss2-esys/esys_crypto_mbed.c

```diff
@@ -528,19 +534,38 @@ iesys_cryptmbed_pk_encrypt(TPM2B_PUBLIC * pub_tpm_key,
     TSS2_RC r = TSS2_RC_SUCCESS;
 
     mbedtls_rsa_context rsa_context;
+#if MBEDTLS_VERSION_MAJOR > 2
+    mbedtls_rsa_init(&rsa_context);
+#endif
 
     switch(pub_tpm_key->publicArea.nameAlg) {
         case TPM2_ALG_SHA1:
+#if MBEDTLS_VERSION_MAJOR > 2
+            mbedtls_rsa_set_padding(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA1);
+#else
             mbedtls_rsa_init(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA1);
+#endif
```

Comment:
it would be much simpler if we just have the case statements select the right constants, and then after the switch call rsa_init or rsa_init and set_padding.

Reply:
Soomething like this:
```diff
diff --git a/src/tss2-esys/esys_crypto_mbed.c b/src/tss2-esys/esys_crypto_mbed.c
index e85dbbddc116..5b544c0ac807 100644
--- a/src/tss2-esys/esys_crypto_mbed.c
+++ b/src/tss2-esys/esys_crypto_mbed.c
@@ -523,6 +523,7 @@ iesys_cryptmbed_random2b(TPM2B_NONCE * nonce, size_t num_bytes)
  * @retval TSS2_ESYS_RC_BAD_VALUE The algorithm of key is not implemented.
  * @retval TSS2_ESYS_RC_GENERAL_FAILURE The internal crypto engine failed.
  */
+//#define MBEDTLS_VERSION_MAJOR 3
 TSS2_RC
 iesys_cryptmbed_pk_encrypt(TPM2B_PUBLIC * pub_tpm_key,
                            size_t in_size,
@@ -533,45 +534,34 @@ iesys_cryptmbed_pk_encrypt(TPM2B_PUBLIC * pub_tpm_key,
 {
     TSS2_RC r = TSS2_RC_SUCCESS;
 
-    mbedtls_rsa_context rsa_context;
-#if MBEDTLS_VERSION_MAJOR > 2
-    mbedtls_rsa_init(&rsa_context);
-#endif
+    int hash = 0;
 
     switch(pub_tpm_key->publicArea.nameAlg) {
         case TPM2_ALG_SHA1:
-#if MBEDTLS_VERSION_MAJOR > 2
-            mbedtls_rsa_set_padding(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA1);
-#else
-            mbedtls_rsa_init(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA1);
-#endif
+            hash = MBEDTLS_MD_SHA1;
             break;
         case TPM2_ALG_SHA256:
-#if MBEDTLS_VERSION_MAJOR > 2
-            mbedtls_rsa_set_padding(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA256);
-#else
-            mbedtls_rsa_init(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA256);
-#endif
+            hash = MBEDTLS_MD_SHA256;
             break;
         case TPM2_ALG_SHA384:
-#if MBEDTLS_VERSION_MAJOR > 2
-            mbedtls_rsa_set_padding(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA384);
-#else
-            mbedtls_rsa_init(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA384);
-#endif
+            hash = MBEDTLS_MD_SHA384;
             break;
         case TPM2_ALG_SHA512:
-#if MBEDTLS_VERSION_MAJOR > 2
-            mbedtls_rsa_set_padding(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA512);
-#else
-            mbedtls_rsa_init(&rsa_context, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA512);
-#endif
+            hash = MBEDTLS_MD_SHA512;
             break;
         default:
             return_error(TSS2_ESYS_RC_NOT_IMPLEMENTED,
                          "Hash algorithm not supported");
     }
 
+    mbedtls_rsa_context rsa_context;
+#if MBEDTLS_VERSION_MAJOR > 2
+    mbedtls_rsa_init(&rsa_context);
+    mbedtls_rsa_set_padding(&rsa_context, MBEDTLS_RSA_PKCS_V21, hash);
+#else
+    mbedtls_rsa_init(&rsa_context, MBEDTLS_RSA_PKCS_V21, hash);
+#endif
+
     if (pub_tpm_key->publicArea.unique.rsa.size == 0) {
         goto_error(r, TSS2_ESYS_RC_BAD_VALUE,
                    "Public key size may not be 0", cleanup);
```

Reply:
not tested....

Reply:
Yes, that's a lot simpler. Init/set padding functions are void, so there's no need for error checking